src/GraphBLAS-sharp.Backend/Vector/SpMSpV.fs

```diff
@@ -0,0 +1,307 @@
+                                    resultRows
+                                    resultIndices
+                                    resultValues)
+                    )
+
+                    queue.Post(Msg.CreateRunMsg<_, _>(gather))
+
+                collectedRows.Free queue
+
+                Some(resultRows, resultIndices, resultValues)
+
+
+    let private multiplyScalar (clContext: ClContext) (mul: Expr<'a option -> 'b option -> 'c option>) workGroupSize =
+
+        let multiply =
+            <@ fun (ndRange: Range1D) resultLength vectorLength (rowIndices: ClArray<int>) (matrixValues: ClArray<'a>) (vectorIndices: ClArray<int>) (vectorValues: ClArray<'b>) (resultValues: ClArray<'c option>) ->
+                let i = ndRange.GlobalID0
+
+                if i < resultLength then
+                    let index = rowIndices.[i]
+                    let matrixValue = matrixValues.[i]
+
+                    let vectorValue =
+                        (%Search.Bin.byKey) vectorLength index vectorIndices vectorValues
+
+                    let res = (%mul) (Some matrixValue) vectorValue
+                    resultValues.[i] <- res @>
+
+        let multiply = clContext.Compile multiply
+
+        fun (queue: MailboxProcessor<_>) (columnIndices: ClArray<int>) (matrixValues: ClArray<'a>) (vector: Sparse<'b>) ->
+
+            let resultLength = columnIndices.Length
+
+            let result =
+                clContext.CreateClArrayWithSpecificAllocationMode(DeviceOnly, resultLength)
+
+            let ndRange =
+                Range1D.CreateValid(resultLength, workGroupSize)
+
+            let multiply = multiply.GetKernel()
+
+            queue.Post(
+                Msg.MsgSetArguments
+                    (fun () ->
+                        multiply.KernelFunc
+                            ndRange
+                            resultLength
+                            vector.NNZ
+                            columnIndices
+                            matrixValues
+                            vector.Indices
+                            vector.Values
+                            result)
+            )
+
+            queue.Post(Msg.CreateRunMsg<_, _>(multiply))
+
+            result
+
+    let run
+        (add: Expr<'c option -> 'c option -> 'c option>)
+        (mul: Expr<'a option -> 'b option -> 'c option>)
+        (clContext: ClContext)
+        workGroupSize
+        =
+
+        //TODO: Common.Gather?
+        let gather = gather clContext workGroupSize
+
+        //TODO: Radix sort
+        let sort =
+            Sort.Bitonic.sortKeyValuesInplace clContext workGroupSize
+
+        let multiplyScalar =
+            multiplyScalar clContext mul workGroupSize
+
+        let segReduce =
+            Reduce.ByKey.Option.segmentSequential add clContext workGroupSize
+
+        fun (queue: MailboxProcessor<_>) (matrix: ClMatrix.CSR<'a>) (vector: ClVector.Sparse<'b>) ->
+
+            match gather queue matrix vector with
```

Comment:
In my opinion, it is better to use `Option.map` or `Option.bind` here. How it is done in `SpGeMM.Expand`